src/msg.c

```diff
@@ -0,0 +1,721 @@
+                        if (printer->unkArraySize == printer->currentLine) {
+                            printer->windowState = 5;
+                            printer->rewindArrowAnimState = 0;
+                            printer->rewindArrowBlinkCounter = 0;
+                        }
+                    }
+                } else {
+                    printer->curLinePos -= phi_a0_5;
+                    if (printer->lineEndPos[printer->unkArraySize] >= printer->curLinePos) {
+                        printer->curLinePos = printer->lineEndPos[printer->unkArraySize];
+                        printer->windowState = 0xC;
+                    }
+                }
+            default:
+            block_135:
+                if (printer->stateFlags & 1) {
+                    printer->windowState = 0;
+                    printer->stateFlags = 0;
+
+                    if (printer->letterBackgroundImg != NULL) {
+                        general_heap_free(printer->letterBackgroundImg);
+                    }
+
+                    if (printer->letterBackgroundPal != NULL) {
+                        general_heap_free(printer->letterBackgroundPal);
+                    }
+
+                    if (printer->letterContentImg != NULL) {
+                        general_heap_free(printer->letterContentImg);
+                    }
+
+                    if (printer->letterContentPal != NULL) {
+                        general_heap_free(printer->letterContentPal);
+                    }
+
+                    if (printer->closedWritebackBool != NULL) {
+                        *printer->closedWritebackBool = 1;
+                        printer->closedWritebackBool = NULL;
+                    }
+                }
+        }
+    }
+    return printer->windowState;
+}
+#else
+INCLUDE_ASM(s32, "msg", _update_message, MessagePrintState* msgPrintState);
+#endif
+
+INCLUDE_ASM(s32, "msg", render_messages);
+
+void msg_play_speech_sound(MessagePrintState* printer, u8 character) {
+    f32 volTemp;
+    s16 volume;
+    s32 pitchShift;
+    s32 flag = 1;
+    s32 hundred = 100;
+
+    if (printer->stateFlags & 0x800000 && !(printer->delayFlags & 6) && printer->volume != 0) {
+        volTemp = (f32)printer->volume / 100.0;
+        pitchShift = ((character % 20) * 10) + (printer->speechVolumePitch - hundred);
+        volume = ((rand_int(15) + 78) * volTemp);
+
+        if (volume > 255) {
+            volume = 255;
+        }
+
+        if (character & flag) {
+            sfx_play_sound_with_params(printer->speedSoundIDA, volume, printer->speechPan, pitchShift);
+        } else {
+            sfx_play_sound_with_params(printer->speedSoundIDB, volume, printer->speechPan, pitchShift);
+        }
+    }
+}
+
+INCLUDE_ASM(s32, "msg", msg_copy_to_print_buffer);
+
+void initialize_printer(MessagePrintState* printer, s32 arg1, s32 arg2) {
+    s32 i;
+
+    printer->printBufferSize = ARRAY_COUNT(printer->printBuffer);
+    printer->printBuffer[0] = 0xFB;
```

Comment:
Worth noting - possibly through a comment? - that this is `[next]`; we should probably make an enum for all the characters/opcodes (see tools/build/msg/parse_compile.py).